### PR TITLE
[ODS-6000] Add MultiTenant with context routes example

### PR DESCRIPTION
**Pull request by semalaiappan**
try this application_name with '2023' & '2022' in tenant1 & tenant2 in db admin 
DO $$ 
DECLARE
vendor_name varchar(150) := 'Local Test Vendor';
namespace_prefix varchar(255) = 'uri://ed-fi.org';
user_full_name varchar(150) = 'Local Test User';
user_email_address varchar(150) = 'localtest@ed-fi.org';
application_name varchar(255) = '2023';
claimset_name varchar(255) = 'Ed-Fi Sandbox';
api_client_name varchar(50) = 'Local Test Api Client';
education_organization_id int = 255901; --Must be an ed-org in the ODS
client_key varchar(50) = 'testkey';
client_secret varchar(100) = 'testsecret';
is_populated_sandbox int = 1;

-- For Non-Sandbox deployments
use_sandbox boolean = False;
-- For Sandbox
--use_sandbox boolean = 1;
vendor_id int;
user_id int;
application_id int;
application_education_organization_id int;
api_client_id int;

BEGIN 

-- Clear is_populated_sandbox if not using sandbox
IF NOT use_sandbox
THEN
    SELECT 0 INTO is_populated_sandbox;
END IF;

-- Ensure Vendor exists
SELECT VendorId INTO vendor_id FROM dbo.Vendors WHERE VendorName = vendor_name;
IF(vendor_id IS NULL)
THEN
    INSERT INTO dbo.Vendors (VendorName)
    VALUES (vendor_name);
    SELECT LASTVAL() INTO vendor_id;
END IF;

-- Ensure correct namespace prefixes are set up
DELETE FROM dbo.VendorNamespacePrefixes WHERE Vendor_VendorId = vendor_id;
INSERT INTO dbo.VendorNamespacePrefixes (Vendor_VendorId, NamespacePrefix)
VALUES (vendor_id, namespace_prefix);

-- Ensure User exists for test Vendor
SELECT UserId INTO user_id FROM dbo.Users WHERE FullName = user_full_name AND Vendor_VendorId = vendor_id;
IF(user_id IS NULL)
THEN
    INSERT INTO dbo.Users (Email, FullName, Vendor_VendorId)
    VALUES (user_email_address, user_full_name, vendor_id);
    SELECT LASTVAL() INTO user_id;
ELSE
    UPDATE dbo.Users SET Email = user_email_address WHERE UserId = user_id;
END IF;

-- Ensure Application exists
SELECT ApplicationId INTO application_id FROM dbo.Applications WHERE ApplicationName = application_name AND Vendor_VendorId = vendor_id;
IF (application_id IS NULL)
THEN 
    INSERT INTO dbo.Applications (ApplicationName, Vendor_VendorId, ClaimSetName)
    VALUES (application_name, vendor_id, claimset_name);
    SELECT LASTVAL() INTO application_id;
ELSE
    UPDATE dbo.Applications SET ClaimSetName = claimset_name WHERE ApplicationId = application_id;
END IF;

-- Ensure ApiClient exists
SELECT  ApiClientId INTO api_client_id FROM dbo.ApiClients WHERE Application_ApplicationId = application_id AND Name = api_client_name;
IF(api_client_id IS NULL)
THEN
    INSERT INTO dbo.ApiClients (Key, Secret, Name, IsApproved, UseSandbox, SandboxType, Application_ApplicationId, User_UserId, SecretIsHashed)
    VALUES (client_key, client_secret, api_client_name, TRUE, use_sandbox, is_populated_sandbox, application_id, user_id, FALSE);
    SELECT  LASTVAL() INTO api_client_id;
ELSE
    UPDATE dbo.ApiClients SET Key = client_key, Secret = client_secret, UseSandbox = use_sandbox, SandboxType = is_populated_sandbox, User_UserId = user_id, SecretIsHashed = FALSE WHERE ApiClientId = api_client_id;
END IF;

-- Ensure ApiClientOdsInstance exists
INSERT INTO dbo.ApiClientOdsInstances (ApiClient_ApiClientId, OdsInstance_OdsInstanceId)
SELECT api_client_id, odsinstanceid FROM dbo.OdsInstances;

IF (education_organization_id IS NOT NULL)
THEN
    -- Clear all education organization links for the selected application
    DELETE 
    FROM dbo.ApiClientApplicationEducationOrganizations WHERE 
    ApplicationEdOrg_ApplicationEdOrgId IN ( SELECT ApplicationEducationOrganizationId 
                                             FROM dbo.ApplicationEducationOrganizations 
                                             WHERE Application_ApplicationId = application_id);
    DELETE FROM dbo.ApplicationEducationOrganizations WHERE Application_ApplicationId = application_id;
    -- Ensure correct education organizations are set up
    INSERT INTO dbo.ApplicationEducationOrganizations (EducationOrganizationId, Application_ApplicationId)
    VALUES (education_organization_id, application_id);
    SELECT  LASTVAL() INTO application_education_organization_id;
    INSERT INTO dbo.ApiClientApplicationEducationOrganizations (Applicationedorg_Applicationedorgid, ApiClient_ApiClientId)
    VALUES (application_education_organization_id, api_client_id);
END IF;
END $$;

### CODE REVIEW

**Comment by semalaiappan** [comment]
docker build --pull --rm -f Compose-Generator/Alpine/Dockerfile -t edfiodsdocker:latest Compose-Generator/Alpine

docker run --rm -v C:/OSS-Workspace/Ed-Fi-ODS-Docker/Compose-Generator/Alpine/templates/pgsql/MultiTenant-OdsContext/parameters.example.yml:/parameters -v C:/OSS-Workspace/Ed-Fi-ODS-Docker/Compose/pgsql/MultiTenant-OdsContext:/output edfiodsdocker:latest MultiTenant-OdsContext.sh

 .\multi-tenant-odscontexts-clean.ps1   

.\multi-tenant-odscontexts-up.ps1

File: Compose-Generator/Alpine/Dockerfile [hunk 1]
```diff
@@ -6,7 +6,7 @@
 # Tag alpine:3.18.3
 FROM alpine:3.18.3@sha256:7144f7bab3d4c2648d7e59409f15ec52a18006a128c733fcff20d3a4a54ba44a
 LABEL maintainer="Ed-Fi Alliance, LLC and Contributors <techsupport@ed-fi.org>"
-
+RUN apk --no-cache add dos2unix=~7
```
Comment (simpat-jesus) [inline Compose-Generator/Alpine/Dockerfile:9]: I tested without this line and it worked fine, Alpine 3.18.3 image already has dos2unix installed

File: Compose-Generator/Alpine/templates/pgsql/MultiTenant-OdsContext/run.sh [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+# Licensed to the Ed-Fi Alliance under one or more agreements.
+# The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
+# See the LICENSE and NOTICES files in the project root for more information.
+
+/app/mustache parameters /app/multiTenant-OdsContext/appsettings-template.mustache > output/appsettings.dockertemplate.json
+/app/mustache parameters /app/multiTenant-OdsContext/bootstrap-template.mustache > output/bootstrap.sh
+/app/mustache parameters /app/multiTenant-OdsContext/compose-template.mustache > output/compose-multi-tenant-odscontext-env.yml
+
+dos2unix /output/*.sh
+chmod 777 /output/bootstrap.sh
```
Comment (simpat-jesus) [inline Compose-Generator/Alpine/templates/pgsql/MultiTenant-OdsContext/run.sh:12]: use chmod 755, also add both dos2unix and chmod to Multitenant and SingleTenant-OdsContext run.sh scripts

File: Compose-Generator/Alpine/templates/pgsql/MultiTenant-OdsContext/compose-template.mustache [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+# Licensed to the Ed-Fi Alliance under one or more agreements.
+# The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
+# See the LICENSE and NOTICES files in the project root for more information.
+
+version: "3.8"
+
+services:
+{{#tenants}}
+{{#tokens}}
+
```
Comment (simpat-jesus) [inline Compose-Generator/Alpine/templates/pgsql/MultiTenant-OdsContext/compose-template.mustache:11]: Not a breaking issue, but we can format the template to only have empty spaces before a {{/ }} line

**Review by simpat-jesus — CHANGES_REQUESTED** [review]
Some minor changes

File: Compose-Generator/Alpine/Dockerfile (hunk 1 above)
Comment (semalaiappan) [reply Compose-Generator/Alpine/Dockerfile]: Removed it . Thanks

File: Compose-Generator/Alpine/templates/pgsql/MultiTenant-OdsContext/run.sh (hunk 2 above)
Comment (semalaiappan) [reply Compose-Generator/Alpine/templates/pgsql/MultiTenant-OdsContext/run.sh]: Updated it . Thanks

File: Compose-Generator/Alpine/templates/pgsql/MultiTenant-OdsContext/compose-template.mustache (hunk 3 above)
Comment (semalaiappan) [reply Compose-Generator/Alpine/templates/pgsql/MultiTenant-OdsContext/compose-template.mustache]: Formatted it . Thanks

**Review by simpat-jesus — APPROVED** [review]

**Pull request merged**